File: kotlin/src/03.kt
fun main() {
    val ls = readInput()

    fun part1(): Int {
        var ans = 0

        val pattern = Regex("mul\\((\\d{1,3}),(\\d{1,3})\\)")

        for (l in ls) {
            for (match in pattern.findAll(l)) {
                val (_, a, b) = match.groupValues
                ans += a.toInt() * b.toInt()
            }
        }

        return ans
    }

    fun part2(): Int {
        var ans = 0

        val pattern = Regex("mul\\((\\d{1,3}),(\\d{1,3})\\)|(do\\(\\))|(don't\\(\\))")

        var mult = true
        for (l in ls) {
            for (match in pattern.findAll(l)) {
                val (_, a, b, do_, dont) = match.groupValues
                if (do_ != "")
                    mult = true
                else if (dont != "")
                    mult = false
                else if (mult)
                    ans += a.toInt() * b.toInt()
            }
        }

        return ans
    }

    println(part1())
    println(part2())
}

File: 03.py
from aoc import *

DAY = 3

s = get_input(DAY)
#s = get_example(DAY)

import collections
import math
import re

ls = s.splitlines()


def part1():
    ans = 0

    pattern = "mul\\((\\d{1,3}),(\\d{1,3})\\)"

    for l in ls:
        for a, b in re.findall(pattern, l):
            ans += int(a) * int(b)

    return ans


def part2():
    ans = 0

    pattern = "mul\\((\\d{1,3}),(\\d{1,3})\\)|(do\\(\\))|(don't\\(\\))"

    mult = True
    for l in ls:
        for a, b, do, dont in re.findall(pattern, l):
            if do:
                mult = True
            elif dont:
                mult = False
            elif mult:
                ans += int(a) * int(b)

    return ans


# submit(DAY, part1(), 1)
# submit(DAY, part2(), 2)
